로또 게임 구현

Gobukgol opened the pull request.

jyami-kim committed:
team2's source

Gobukgol committed:
로또 게임 구현
1. domain ->
	1. Lotto : 임의의 6자리 숫자 리스트를 가지고있음
	2. LottoGame : 금액에 맞춰서 로또 리스트를 생성하고 LottoResult를 반환
	3. LottoResult : 로또게임 결과를 관리하는 dto
2. util ->
	1. FixedListGeneratorImpl : 전달한 시작값부터 6개 숫자리스트 반환
	2. RandomListGeneratorImpl : 랜덤한 6개 숫자리스트 반환
	3. ReceivedType : 맞춘숫자에 따른 수령금액을 Enum클래스로 관리
3. view ->
	1. 입력 및 출력

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,40 @@
+        return answer;
+    }
+
+    public boolean isWinNumbers(List<Integer> correctAnswer){
```
settter인거같다

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,40 @@
+        return answer > 2;
+    }
+
+    public String convert(){
```
toString으로 호출하는게 더 좋을듯

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,40 @@
+                .filter(correctAnswer::contains)
+                .count();
+
+        return answer > 2;
```
상수로 빼면 좀더 명확할듯

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.stream.Collectors;
+
+public class LottoResult {
+    public static final int LOTTO_PRICE = 1000;
```
로또결과를 보여주는클래스보다는 로또 게임에 있는게 더 좋을거같아

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/util/ReceivedType.java:
```diff
@@ -0,0 +1,38 @@
+                .receivedAmount;
+    }
+
+    public boolean findByCount(int answerCount){
```
find보단 다른 함수이름이 좋을거같아

Gobukgol committed:
1.Lotto의 번호 출력 방식 변경
	convert() -> toString()
2.Lotto클래스의 convert함수 제거
3.Lotto의 번호리스트를 생성자에서 정렬하는것으로 변경

Gobukgol committed:
1.LottoResult의 LOTTO_PRICE 상수 위치 변경
	LottoResult -> LottoGame
2.ReceivedType클래스의 findByCount함수명 변경
	findByCount -> isCorrectAnswer

Gobukgol committed:
1. Lotto의 getAnswer함수 변경 및 isWinnigCount함수 제거
2. LottoResult에서 정답 번호 리스트를 가지고있도록 변경
3. LottoGameTest에서 로또 리스트 생성 테스트 추가

pci2676 commented on src/main/java/com/ccstudy/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,55 @@
+    public int getReceivedAmount() {
+        int receivedAmount = 0;
+
+        for(Lotto lotto : winningLottos){
```
이 부분 스트림으로 풀어서 할 수 있지 않을까?

pci2676 commented on src/main/java/com/ccstudy/lotto/util/ReceivedType.java:
```diff
@@ -0,0 +1,38 @@
+                .filter(receivedType -> receivedType.isCorrectCount(answerCount))
+                .findAny()
+                .get()
+                .receivedAmount;
```
필터에 걸린거 하나도 없으면 0 리턴되는거야????

Gobukgol committed:
19.03.23 로또 코드리뷰에 따른 코드 수정

1. NumberListGenerator 클래스 및 correctAnswer 변수 네이밍 변경
2. LottoResult Yield계산 부분 변경

Gobukgol committed:
19.03.23 로또 코드리뷰에 따른 코드 수정(2)

1. 테스트 코드의 반복되는 로직 함수로 처리

Gobukgol committed:
1.ReceivedType.class 이름 변경
     ReceivedType -> LottoRank;
2.당첨번호와 보너스 번호를 관리하는 WinningNumber 클래스 추가
     LottoGame,LottoResult의 생성자 매개변수 변경
3.수동 로또와 보너스 볼 기능 추가에 따른 로직 변경
     Input, Output, LottoGame, LottoResult, LottoRank 등 변경
3.Input클래스 테스트 코드 생성

Gobukgol committed:
1. 입력받은 값을 가지고 있는 InputDto 클래스 생성
   -> 수동 입력 갯수와, 수동 번호를 String으로 가지고 있음
2. 로또의 번호 int를 감싸는 LottoNo 클래스 생성
   -> 1~45 범위의 값이 맞지 않으면 RunTimeException 발생
3. LottoRank의 correctBonus 멤버변수 제거 및 getRank함수 제거
   -> LottoRank의 함수를 통해 Rank를 가지고 오지않고 LottoResult에서 반복을 돌며
      가져옴
4. Lotto 클래스 이름 변경
   -> Lotto -> LottoTicket
5. LottoMachine 클래스 생성
   -> 자동, 수동 로또에 맞게 로또 리스트 반환

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,41 @@
+
+        List<String> manualLottoTickets = input.inputManualLottoNumber(manualAmountOfLotto);
+
+        InputDto inputDto = new InputDto(manualAmountOfLotto, manualLottoTickets);
```
인풋 클래스에서 한번에 inputDto를 반환하도록하면 좋을거같아 지금 main에 너무 몰빵되어있어

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,37 @@
+        return lottoTickets;
+    }
+
+    public LottoResult gameStart(WinningNumber winningNumber) {
```
게임스타트인데 실제로 결과 생성만 생성하니까 의미가 이상한거같아

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,73 @@
+
+public class LottoMachine {
+    private static final String SPLITTER = ",";
+    private static List<Integer> lottoRange = IntStream.range(1, 45)
```
이것도 LottoNo으로 만들면 좋을듯

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,73 @@
+        List<LottoTicket> fixedLottos = new ArrayList<>();
+
+        for (int i = 0; i < amountOfFixedLotto; i++) {
+            List<Integer> sixList;
```
여섯개의 리스트?

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,73 @@
+        return manualLottos;
+    }
+
+    public static List<LottoTicket> createFixedLottoTicket(int amountOfFixedLotto) {
```
열줄 넘는 함수

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,76 @@
+    public LottoRank getMatchingRank(LottoTicket lottoTicket) {
+        LottoRank matchingRank = LottoRank.DEFAULT;
+
+        for (LottoRank rank : LottoRank.values()) {
```
인덴트 두번 빠진거 빼면좋을거같은데 이등을 검사하는 로직을 넣으면 될듯

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/view/Input.java:
```diff
@@ -0,0 +1,64 @@
+
+    public int inputManualAmountOfLotto() {
+        System.out.println("수동으로 구매할 로또 수를 입력해주세요.");
+        return Integer.parseInt(scanner.nextLine());
```
nextInt로 받으면 되지않나?

chldbtjd2272 commented on src/main/java/com/ccstudy/lotto/domain/LottoTicket.java:
```diff
@@ -0,0 +1,27 @@
+        this.lottoNumber = lottoNumber;
+    }
+
+    public List<Integer> getLottoNumber() {
```
s

chldbtjd2272 commented on src/test/java/com/ccstudy/lotto/domain/LottoMachineTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Test
+    public void 수동_로또_생성_테스트() {
+        List<String> manualNumbers = new ArrayList<>();
```
Arrays.asList로 고치면 좋을듯

Gobukgol committed:
1. input 클래스에서 inputDto와 WinningNumber를 반환하도록 변경
   -> main에 들어가는 로직을 줄임
   -> inputDto는 구매금액,수동 로또 개수, 수동 번호를 가지고있음
2. LottoNo에서 validation 체크 로직 변경
3. LottoMachine 클래스의 getSixLottoNoList함수 제거
4. LottoTicket 클래스의 lottoNumber 멤버변수 이름 변경
5. LottoResult 클래스 getMatchinRank함수 로직 변경
   -> 인덴트가 두번 들어가있는것을 stream을 이용하여 제거
